### ISSUE TITLE
No method to get the file downloading/uploading progress

### ISSUE THREAD

**Issue by thekindJose** [dropbox/dropbox-sdk-java#66]
Hi, 
I'm trying to integrate the v2 dropbox sdk into my android app. 
I'm working on downloading files, but I cannot find anything in the API which notifies me regarding the downloading progress (as there was before on v1.6, i.e. the ProgressListener). 
Am I missing something or have you removed this important feature?

Thanks!

**Comment by greg-db** [comment]
The API v2 Java SDK doesn't offer progress listeners, but we'll consider this a feature request. (The v2 Java SDK isn't built on the v1 Android Core SDK, so it didn't inherit the progress listeners from there.)

**Comment by thekindJose** [comment]
Hi @greg-db. Is it possible to know when could we expect it to be done? We would like to migrate to v2 on our android app, but we do not want to lose functionality. If user is downloading a file, user wants to know how is the progress of the download going. Even in Java I think it is an important thing to have, isn't it? I'm not talking of a ProgressListener android specific, but at least a callback to get notified. Looking at the code seems to be closed for extension.

**Comment by greg-db** [comment]
I don't have a timeline to offer for this unfortunately. Apologies I don't have any more useful to share!

**Comment by thekindJose** [comment]
@greg-db, may I know which version of the SDK you are using for the android dropbox app? I've got installed on my phone dropbox version 15.2.2 and on this apk there is a progressbar both for the downloads and for the uploads. Are you using a different API?

**Comment by cakoose** [comment]
@thekindJose: As a workaround, you can create a wrapper `OutputStream` that tracks progress.  Something like:

```
public class ProgressOutputStream extends OutputStream {
    ...
    public ProgressOutputStream(long totalSize, OutputStream underlying, Listener listener) {
        this.underlying = underlying;
        this.listener = listener;
        this.completed = 0;
        this.totalSize = totalSize;
    }

    @Override
    public void write(byte[] data, int off, int len) throws IOException {
        this.underlying.write(data, off, len);
        track(len);
    }

    @Override
    public void write(byte[] data) throws IOException {
        this.underlying.write(data);
        track(data.length);
    }

    @Override
    public void write(int c) {
        this.underlying.write(c);
        track(1)
    }

    private void track(int len) {
        this.completed += len;
        this.listener.progress(this.completed, this.totalSize);
    }

    public interface Listener {
        public void progress(long completed, long totalSize);
    }
}
```

Then, with the SDK:

```
DbxClientV2 client = ...
DbxDownloader<FileMetadata> dl = client.files().download("/remote-file-path");
long size = dl.getResult().size;

FileOutputStream fout = new FileOutputStream("local-file");
dl.download(new ProgressOutputStream(size, fout, new ProgressOutputStream.Listener() {
    void progress(long completed, long totalSize) {
        // update progress bar here ...
    }
});

**Comment by LitbLeo** [comment]
Hi @cakoose, i used your method,it was just worked  by downloading.When i used InputStream to resovle the upload,it doesn't work.Code like this:
`public class ProgressAwareInputStream extends InputStream {
    private InputStream wrappedInputStream;
    private long size;
    private long counter;
    private InputStreamListener listener;

    public ProgressAwareInputStream(InputStream in, long size, InputStreamListener listener) {
        wrappedInputStream = in;
        this.size = size;
        this.listener = listener;
    }
    

    @Override
    public int read() throws IOException {
        int retVal = wrappedInputStream.read();
        counter += 1;
        check(retVal);
        return retVal;
    }

    @Override
    public int read(byte[] b) throws IOException {
        int retVal = wrappedInputStream.read(b);
        counter += retVal;
        check(retVal);
        return retVal;
    }

    @Override
    public int read(byte[] b, int offset, int length) throws IOException {
        int retVal = wrappedInputStream.read(b, offset, length);
        counter += retVal;
        check(retVal);
        return retVal;
    }

    private void check(int retVal) {
        if (retVal != -1) {
            listener.uploadProgress(counter, size);
        }
    }

    @Override
    public void close() throws IOException {
        wrappedInputStream.close();
    }
    @Override
    public int available() throws IOException {
        return wrappedInputStream.available();
    }
    @Override
    public void mark(int readlimit) {
        wrappedInputStream.mark(readlimit);
    }
    @Override
    public synchronized void reset() throws IOException {
        wrappedInputStream.reset();
    }
    @Override
    public boolean markSupported() {
        return wrappedInputStream.markSupported();
    }
    @Override
    public long skip(long n) throws IOException {
        return wrappedInputStream.skip(n);
    }

    /**
     * Interface for classes that want to monitor this input stream
     */
    public interface InputStreamListener{
        void uploadProgress(long completed, long totalSize);
    }
}`
Could you help me ?Thanks!

**Comment by wangbourne** [comment]
+1. to show uploading progress is quite a basic and required function, hope this function can be added as soon as possible

**Comment by SDPrio** [comment]
+1 I am also surprised that such an essential feature is not included.

**Comment by nzala** [comment]
an essential feature is not included.

**Comment by tactoth** [comment]
And you just retired the V1 API without this essential feature in the new V2 api you are promoting ;(

**Comment by jiuyangzhao** [comment]
@cakoose already provided a method to wrap OutputStream and tracking download progress.

For Uploading. In V1 android SDK we provided a convenient class ChunkedUploader. In API V2 we no longer provide it but its actually quite simple to implement your own. You can split you file into small chunks and send them through `/upload_session/start`, `/upload_session/append_v2` and `/upload_session/finish`. Also because you split the file you self it's very simple to keep tracking the uploading progress. This is exactly how we implement upload progress listener in V1.

**Comment by ikzjfr0** [comment]
@jiuyangzhao don't you think this solution is a little complicated, you should provide a friendly API  for upload progress. By the way, there's a bug in your chunked upload: https://github.com/dropbox/dropbox-sdk-java/issues/88

**Comment by tactoth** [comment]
The whole point of using this library is that I don't want to be wrapping all the rest calls.

**Comment by greg-db** [comment]
Thanks for the feedback! I definitely see how it would be more helpful to offer a simplified interface for this.

**Comment by jamesdwlee** [comment]
it's announced the dropbox api v1 will be retired as of 6/28/2017 but still no news to support ProgressMonitor?  Without providing the essential function, how v1 can be retried and dropbox force all apps to move to v2?

**Comment by greg-db** [comment]
I don't have an update on this feature request right now.

Note that we actually just announced a general extension for access to API v1 though, until 9/28/2017:

https://blogs.dropbox.com/developers/2017/06/updated-api-v1-deprecation-timeline/

**Comment by menglingpeng** [comment]
@LitbLeo thank your uploadProgress method,but i think it has a little mistake.
`check(counter);`instead of `check(retVal);`
then it works very well.

**Comment by isabsent** [comment]
Dear Dropbos team! You are forcing developers to migrate on v.2, but you have not provide essential API calls for it! More over, you are writing in your  [announcement](https://blogs.dropbox.com/developers/2017/06/updated-api-v1-deprecation-timeline/) 

> Developers have asked us for more time to finish updating their apps

while developers ask you **to provide in v.2 methods that exist in v.1** during last year! But you do nothing for this! Your announcement should be changed to:

> We are not going to provide painless migration from v.1 to v.2 so you have to implement chunked upload yourself.

The best way in this situation is do not turn off v.1 support  until you are not write consistent v.2 release.

**Comment by greg-db** [comment]
Thanks for the additional feedback! I'm sending it to the team.

**Comment by nzala** [comment]
+1

**Comment by jamesdwlee** [comment]
Any update from DropBox? 9/28 is still the deadline without having any update from DropBox yet?

**Comment by greg-db** [comment]
I don't have an update on this feature request. I'll follow up here if/when I do.

And yes, API v1 is planned for retirement on 9/28/17. You can find the full timeline and information in [the blog post](https://blogs.dropbox.com/developers/2016/06/api-v1-deprecated/).

**Comment by assassinshadow0** [comment]
just use the chunkedUploadFile method from here https://github.com/dropbox/dropbox-sdk-java/blob/7ecc15cf0f51d6ae2ba5cdb334aac2c2f3474b87/examples/upload-file/src/main/java/com/dropbox/core/examples/upload_file/Main.java
I'm playing around with it as we speak.

**Comment by pteale** [comment]
Why the hell did you turn off the v1 API when v2 is clearly incomplete
You cant say your adding upload/download functionality and then not provide progress listeners, only a moron would implement it this way
Get this fixed.

**Comment by dkijkuit** [comment]
Chunked file upload, wow... :-1: 

You just made something trivial much more difficult.

**Comment by saorisato** [comment]
wanna share my 2 cents too, I have been also just started to dig into the API v2 and not be able to have a progress bar is just a big joke.

Moreover your provided Android Example code is just filled with overcomplicated shit code

I still have not figured out what on earth TEST OPEN WITH INTERNAL does do what?
and it crashed when I click on Test preview . . .

Dropbox gonna die soon, Gonna switch to Google Drive

**Comment by greg-db** [comment]
Thanks for the feedback everyone!

@saorisato The "open with" test is just internal functionality for Dropbox's use. You can safely ignore that.

Also, can you [open an issue](https://github.com/dropbox/dropbox-sdk-java/issues/new) with the steps to reproduce and the stack trace for the crash your saw? Thanks in advance!

**Comment by johnnyb15613** [comment]
I cannot believe that over a year after this issue was opened it is still not rectified. You guys do know that you are a leader in cloud storage, yet your API has no way to track the progress of an upload or download.
You deserve whatever downfall ensues, this functionality is as pertinent to file storage as all CRUD operations are. Piss poor, putting it lightly. Even though v1 didnt do it well, it at least did it. Taking critical steps backwards. And not addressing it whatsoever. Booooooooooo

**Comment by nebulabox** [comment]
The team should be shamed.

**Comment by jiuyangzhao** [comment]
We will support this in next release.

**Comment by goxr3plus** [comment]
I will cry from happiness if this is true....

**Comment by goxr3plus** [comment]
Good question.... this feature is awaited 3 years now. I am ready to pay to
be added. ♥

**Comment by goxr3plus** [comment]
@jiuyangzhao :) I see you added it , can we have the release  i love you so much :)

**Comment by greg-db** [comment]
This has been released in [v3.0.9](https://github.com/dropbox/dropbox-sdk-java/releases/tag/v3.0.9). There's an example of using it with with [the `uploadAndFinish` method](https://dropbox.github.io/dropbox-sdk-java/api-docs/v3.0.x/com/dropbox/core/DbxUploader.html#uploadAndFinish-java.io.InputStream-long-com.dropbox.core.util.IOUtil.ProgressListener-) for an upload [here](https://github.com/dropbox/dropbox-sdk-java/blob/master/examples/upload-file/src/main/java/com/dropbox/core/examples/upload_file/Main.java#L50). 

It works the same way with file downloads; [the `download` method](https://dropbox.github.io/dropbox-sdk-java/api-docs/v3.0.x/com/dropbox/core/DbxDownloader.html#download-java.io.OutputStream-com.dropbox.core.util.IOUtil.ProgressListener-) optionally takes a `ProgressListener` parameter the same way.
